File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses);
        vector<int> indeg(numCourses,0);
        for (const auto& p : prerequisites){
            int a = p[0], b = p[1];
            adj[b].push_back(a);
            indeg[a]++;
        }
        queue<int> q;
        for (int c=0; c<numCourses; c++)
            if (indeg[c] == 0) q.push(c);
        int taken = 0;
        while (!q.empty()){
            int cur = q.front(); q.pop();
            taken++;
            for (int other : adj[cur]){
                if (--indeg[other] == 0)
                    q.push(other);
            }
        }
        return taken == numCourses;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj = [[] for _ in range(numCourses)]
        indeg = [0] * numCourses
        for p in prerequisites:
            a,b = p
            adj[b].append(a)
            indeg[a] += 1
        q = deque()
        for c in range(numCourses):
            if indeg[c] == 0:
                q.append(c)
        taken = 0
        while len(q) != 0:
            cur = q.popleft()
            taken += 1
            for other in adj[cur]:
                indeg[other] -= 1
                if indeg[other] == 0:
                    q.append(other)
        return taken == numCourses